# Preparation

In [1]:
import torch
torch.__version__

'2.3.0a0+6ddf5cf85e.nv24.04'

In [2]:
import os
import pandas as pd
import numpy as np
import random
from torch import nn

from torch.utils.data import Dataset, DataLoader
from tqdm import tqdm

from model import *
from dataset import *
from train import *

In [3]:
device = (
    "cuda"
    if torch.cuda.is_available()
    else "mps"
    if torch.backends.mps.is_available()
    else "cpu"
)
print(f"Using {device} device")

Using cuda device


In [4]:
BATCH_SIZE = 256
NUMEPOCHS = 50

In [5]:
# import pandas as pd
# df = pd.read_csv("Data/ml-1m.train.rating",sep='\t', header=None, names=['uid', 'mid', 'rating', 'timestamp'], engine='python')
# df

In [6]:
# df2 = pd.read_csv("Data/ml-1m.test.negative",sep = "\t",header=None)
# df2

# Trian and Test

### Compare the three methods 

In [7]:
model_config_mf = {
    "model_mark": "gmf(mf_dim=8)",
    'embedding_dim_mf': 8,
    # "mlp_layers(X)" : 0,
    # 'mlp_layers': [ 32, 16, 8],
    'model_type': 'GMF'     #　MLP, NeuMF
}

model_config_mlp = {
    "model_mark": "mlp(mlp_layer=3)",
    "mlp_layers(X)" : 3,
    'mlp_layers': [32, 16, 8],
    'model_type': 'MLP'     #　MLP, NeuMF
}

model_config_neumf = {
    "model_mark": "neumf(mf_dim=8,mlp_layer=3)",
    'embedding_dim_mf': 8,
    "mlp_layers(X)" : 3,
    'mlp_layers': [32, 16, 8],
    'model_type': 'NeuMF'     #　MLP, NeuMF
}

model_train(model_config_mf, seed = 42, num_of_negatives=4, num_of_epochs=30)


Using cuda device
Loading preprocessed data from preprocessed_data/ml.pkl
config of model: 
 {'model_mark': 'gmf(mf_dim=8)', 'embedding_dim_mf': 8, 'model_type': 'GMF', 'num_users': 6040, 'num_items': 3706}


  3%|▎         | 1/30 [00:57<27:36, 57.12s/it]

Epoch 1, Loss: 0.0027, HR@10: 0.1084, NDCG@10: 0.0498


  7%|▋         | 2/30 [02:00<28:21, 60.77s/it]

Epoch 2, Loss: 0.0020, HR@10: 0.4161, NDCG@10: 0.2293


 10%|█         | 3/30 [03:16<30:29, 67.75s/it]

Epoch 3, Loss: 0.0014, HR@10: 0.4452, NDCG@10: 0.2484


 13%|█▎        | 4/30 [04:11<27:08, 62.63s/it]

Epoch 4, Loss: 0.0014, HR@10: 0.4563, NDCG@10: 0.2525


 17%|█▋        | 5/30 [05:15<26:22, 63.32s/it]

Epoch 5, Loss: 0.0013, HR@10: 0.4724, NDCG@10: 0.2613


 20%|██        | 6/30 [06:38<27:58, 69.92s/it]

Epoch 6, Loss: 0.0013, HR@10: 0.5046, NDCG@10: 0.2770


 23%|██▎       | 7/30 [07:25<23:57, 62.50s/it]

Epoch 7, Loss: 0.0012, HR@10: 0.5334, NDCG@10: 0.2927


 27%|██▋       | 8/30 [08:44<24:47, 67.63s/it]

Epoch 8, Loss: 0.0012, HR@10: 0.5512, NDCG@10: 0.3046


 30%|███       | 9/30 [09:48<23:19, 66.65s/it]

Epoch 9, Loss: 0.0011, HR@10: 0.5699, NDCG@10: 0.3170


 33%|███▎      | 10/30 [10:43<21:01, 63.08s/it]

Epoch 10, Loss: 0.0011, HR@10: 0.5848, NDCG@10: 0.3264


 37%|███▋      | 11/30 [12:07<21:58, 69.42s/it]

Epoch 11, Loss: 0.0011, HR@10: 0.5970, NDCG@10: 0.3356


 40%|████      | 12/30 [13:01<19:25, 64.73s/it]

Epoch 12, Loss: 0.0011, HR@10: 0.6038, NDCG@10: 0.3402


 43%|████▎     | 13/30 [14:16<19:13, 67.84s/it]

Epoch 13, Loss: 0.0011, HR@10: 0.6108, NDCG@10: 0.3456


 47%|████▋     | 14/30 [15:24<18:05, 67.84s/it]

Epoch 14, Loss: 0.0011, HR@10: 0.6147, NDCG@10: 0.3482


 50%|█████     | 15/30 [16:19<15:59, 63.96s/it]

Epoch 15, Loss: 0.0011, HR@10: 0.6179, NDCG@10: 0.3506


 53%|█████▎    | 16/30 [17:20<14:43, 63.10s/it]

Epoch 16, Loss: 0.0011, HR@10: 0.6242, NDCG@10: 0.3546


 57%|█████▋    | 17/30 [18:32<14:15, 65.78s/it]

Epoch 17, Loss: 0.0011, HR@10: 0.6258, NDCG@10: 0.3548


 60%|██████    | 18/30 [19:22<12:11, 60.92s/it]

Epoch 18, Loss: 0.0011, HR@10: 0.6238, NDCG@10: 0.3547


 63%|██████▎   | 19/30 [20:35<11:49, 64.51s/it]

Epoch 19, Loss: 0.0010, HR@10: 0.6275, NDCG@10: 0.3576


 67%|██████▋   | 20/30 [21:36<10:36, 63.68s/it]

Epoch 20, Loss: 0.0010, HR@10: 0.6267, NDCG@10: 0.3586


 70%|███████   | 21/30 [22:30<09:06, 60.74s/it]

Epoch 21, Loss: 0.0010, HR@10: 0.6275, NDCG@10: 0.3585


 73%|███████▎  | 22/30 [23:55<09:02, 67.80s/it]

Epoch 22, Loss: 0.0010, HR@10: 0.6255, NDCG@10: 0.3590


 77%|███████▋  | 23/30 [24:48<07:24, 63.57s/it]

Epoch 23, Loss: 0.0010, HR@10: 0.6298, NDCG@10: 0.3604


 80%|████████  | 24/30 [26:01<06:37, 66.17s/it]

Epoch 24, Loss: 0.0010, HR@10: 0.6301, NDCG@10: 0.3620


 83%|████████▎ | 25/30 [27:10<05:35, 67.09s/it]

Epoch 25, Loss: 0.0010, HR@10: 0.6295, NDCG@10: 0.3610


 87%|████████▋ | 26/30 [28:11<04:21, 65.30s/it]

Epoch 26, Loss: 0.0010, HR@10: 0.6301, NDCG@10: 0.3617


 90%|█████████ | 27/30 [29:23<03:21, 67.22s/it]

Epoch 27, Loss: 0.0010, HR@10: 0.6311, NDCG@10: 0.3622


 93%|█████████▎| 28/30 [30:29<02:14, 67.04s/it]

Epoch 28, Loss: 0.0010, HR@10: 0.6298, NDCG@10: 0.3614


 97%|█████████▋| 29/30 [31:24<01:03, 63.45s/it]

Epoch 29, Loss: 0.0010, HR@10: 0.6303, NDCG@10: 0.3611


100%|██████████| 30/30 [32:49<00:00, 65.65s/it]

Epoch 30, Loss: 0.0010, HR@10: 0.6311, NDCG@10: 0.3611
Best HR: 0.6311258278145695, Best NDCG: 0.3622338343922371 at Epoch 26


In [8]:
model_train(model_config_mlp, seed = 42, num_of_negatives=4, num_of_epochs=30)


Using cuda device
Loading preprocessed data from preprocessed_data/ml.pkl
config of model: 
 {'model_mark': 'mlp(mlp_layer=3)', 'mlp_layers(X)': 3, 'mlp_layers': [32, 16, 8], 'model_type': 'MLP', 'num_users': 6040, 'num_items': 3706}


  3%|▎         | 1/30 [01:57<56:58, 117.90s/it]

Epoch 1, Loss: 0.0015, HR@10: 0.4434, NDCG@10: 0.2497


  7%|▋         | 2/30 [03:22<45:59, 98.56s/it] 

Epoch 2, Loss: 0.0014, HR@10: 0.4515, NDCG@10: 0.2518


 10%|█         | 3/30 [05:03<44:45, 99.46s/it]

Epoch 3, Loss: 0.0014, HR@10: 0.4545, NDCG@10: 0.2537


 13%|█▎        | 4/30 [06:37<42:07, 97.22s/it]

Epoch 4, Loss: 0.0014, HR@10: 0.4624, NDCG@10: 0.2574


 17%|█▋        | 5/30 [08:06<39:16, 94.27s/it]

Epoch 5, Loss: 0.0013, HR@10: 0.4829, NDCG@10: 0.2651


 20%|██        | 6/30 [09:38<37:25, 93.54s/it]

Epoch 6, Loss: 0.0013, HR@10: 0.4911, NDCG@10: 0.2717


 23%|██▎       | 7/30 [11:29<38:03, 99.29s/it]

Epoch 7, Loss: 0.0013, HR@10: 0.5005, NDCG@10: 0.2741


 27%|██▋       | 8/30 [13:05<36:04, 98.37s/it]

Epoch 8, Loss: 0.0013, HR@10: 0.5070, NDCG@10: 0.2787


 30%|███       | 9/30 [14:48<34:53, 99.71s/it]

Epoch 9, Loss: 0.0013, HR@10: 0.5132, NDCG@10: 0.2842


 33%|███▎      | 10/30 [16:29<33:20, 100.02s/it]

Epoch 10, Loss: 0.0012, HR@10: 0.5129, NDCG@10: 0.2839


 37%|███▋      | 11/30 [17:58<30:37, 96.69s/it] 

Epoch 11, Loss: 0.0012, HR@10: 0.5242, NDCG@10: 0.2899


 40%|████      | 12/30 [19:40<29:30, 98.39s/it]

Epoch 12, Loss: 0.0012, HR@10: 0.5263, NDCG@10: 0.2924


 43%|████▎     | 13/30 [21:14<27:29, 97.03s/it]

Epoch 13, Loss: 0.0012, HR@10: 0.5275, NDCG@10: 0.2976


 47%|████▋     | 14/30 [22:41<25:04, 94.04s/it]

Epoch 14, Loss: 0.0012, HR@10: 0.5301, NDCG@10: 0.2975


 50%|█████     | 15/30 [24:18<23:40, 94.71s/it]

Epoch 15, Loss: 0.0012, HR@10: 0.5412, NDCG@10: 0.3031


 53%|█████▎    | 16/30 [25:50<21:58, 94.17s/it]

Epoch 16, Loss: 0.0012, HR@10: 0.5421, NDCG@10: 0.3062


 57%|█████▋    | 17/30 [27:18<19:58, 92.17s/it]

Epoch 17, Loss: 0.0012, HR@10: 0.5439, NDCG@10: 0.3069


 60%|██████    | 18/30 [28:57<18:51, 94.33s/it]

Epoch 18, Loss: 0.0012, HR@10: 0.5478, NDCG@10: 0.3064


 63%|██████▎   | 19/30 [30:20<16:39, 90.86s/it]

Epoch 19, Loss: 0.0012, HR@10: 0.5588, NDCG@10: 0.3112


 67%|██████▋   | 20/30 [31:51<15:07, 90.80s/it]

Epoch 20, Loss: 0.0012, HR@10: 0.5533, NDCG@10: 0.3090


 70%|███████   | 21/30 [33:24<13:42, 91.43s/it]

Epoch 21, Loss: 0.0012, HR@10: 0.5518, NDCG@10: 0.3075


 73%|███████▎  | 22/30 [34:57<12:16, 92.08s/it]

Epoch 22, Loss: 0.0012, HR@10: 0.5579, NDCG@10: 0.3101


 77%|███████▋  | 23/30 [36:22<10:29, 89.94s/it]

Epoch 23, Loss: 0.0012, HR@10: 0.5616, NDCG@10: 0.3128


 80%|████████  | 24/30 [37:55<09:04, 90.79s/it]

Epoch 24, Loss: 0.0012, HR@10: 0.5632, NDCG@10: 0.3139


 83%|████████▎ | 25/30 [39:18<07:23, 88.61s/it]

Epoch 25, Loss: 0.0012, HR@10: 0.5667, NDCG@10: 0.3159


 87%|████████▋ | 26/30 [40:54<06:02, 90.54s/it]

Epoch 26, Loss: 0.0012, HR@10: 0.5637, NDCG@10: 0.3134


 90%|█████████ | 27/30 [42:19<04:27, 89.07s/it]

Epoch 27, Loss: 0.0012, HR@10: 0.5649, NDCG@10: 0.3134


 93%|█████████▎| 28/30 [43:48<02:57, 89.00s/it]

Epoch 28, Loss: 0.0011, HR@10: 0.5644, NDCG@10: 0.3135


 97%|█████████▋| 29/30 [45:14<01:27, 87.99s/it]

Epoch 29, Loss: 0.0011, HR@10: 0.5581, NDCG@10: 0.3146


100%|██████████| 30/30 [46:49<00:00, 93.65s/it]

Epoch 30, Loss: 0.0011, HR@10: 0.5598, NDCG@10: 0.3126
Best HR: 0.5667218543046357, Best NDCG: 0.3158625751324939 at Epoch 24


In [9]:
model_train(model_config_neumf, seed = 42, num_of_negatives=4, num_of_epochs=30)


Using cuda device
Loading preprocessed data from preprocessed_data/ml.pkl
config of model: 
 {'model_mark': 'neumf(mf_dim=8,mlp_layer=3)', 'embedding_dim_mf': 8, 'mlp_layers(X)': 3, 'mlp_layers': [32, 16, 8], 'model_type': 'NeuMF', 'num_users': 6040, 'num_items': 3706}


  3%|▎         | 1/30 [01:34<45:50, 94.85s/it]

Epoch 1, Loss: 0.0015, HR@10: 0.4462, NDCG@10: 0.2497


  7%|▋         | 2/30 [03:20<47:20, 101.44s/it]

Epoch 2, Loss: 0.0014, HR@10: 0.4482, NDCG@10: 0.2512


 10%|█         | 3/30 [05:04<46:07, 102.49s/it]

Epoch 3, Loss: 0.0014, HR@10: 0.4531, NDCG@10: 0.2490


 13%|█▎        | 4/30 [06:56<46:04, 106.31s/it]

Epoch 4, Loss: 0.0014, HR@10: 0.4647, NDCG@10: 0.2561


 17%|█▋        | 5/30 [08:38<43:36, 104.64s/it]

Epoch 5, Loss: 0.0013, HR@10: 0.4876, NDCG@10: 0.2684


 20%|██        | 6/30 [10:23<41:56, 104.87s/it]

Epoch 6, Loss: 0.0012, HR@10: 0.5197, NDCG@10: 0.2869


 23%|██▎       | 7/30 [12:13<40:48, 106.46s/it]

Epoch 7, Loss: 0.0012, HR@10: 0.5450, NDCG@10: 0.3023


 27%|██▋       | 8/30 [13:55<38:27, 104.90s/it]

Epoch 8, Loss: 0.0011, HR@10: 0.5657, NDCG@10: 0.3136


 30%|███       | 9/30 [15:35<36:13, 103.50s/it]

Epoch 9, Loss: 0.0011, HR@10: 0.5826, NDCG@10: 0.3252


 33%|███▎      | 10/30 [17:09<33:33, 100.69s/it]

Epoch 10, Loss: 0.0011, HR@10: 0.5959, NDCG@10: 0.3341


 37%|███▋      | 11/30 [18:58<32:41, 103.22s/it]

Epoch 11, Loss: 0.0010, HR@10: 0.6104, NDCG@10: 0.3450


 40%|████      | 12/30 [20:23<29:14, 97.45s/it] 

Epoch 12, Loss: 0.0010, HR@10: 0.6172, NDCG@10: 0.3482


 43%|████▎     | 13/30 [21:55<27:12, 96.03s/it]

Epoch 13, Loss: 0.0010, HR@10: 0.6228, NDCG@10: 0.3535


 47%|████▋     | 14/30 [23:12<24:02, 90.16s/it]

Epoch 14, Loss: 0.0010, HR@10: 0.6300, NDCG@10: 0.3560


 50%|█████     | 15/30 [24:48<23:00, 92.04s/it]

Epoch 15, Loss: 0.0010, HR@10: 0.6321, NDCG@10: 0.3584


 53%|█████▎    | 16/30 [26:25<21:46, 93.35s/it]

Epoch 16, Loss: 0.0010, HR@10: 0.6341, NDCG@10: 0.3605


 57%|█████▋    | 17/30 [28:00<20:19, 93.82s/it]

Epoch 17, Loss: 0.0010, HR@10: 0.6356, NDCG@10: 0.3624


 60%|██████    | 18/30 [29:40<19:08, 95.73s/it]

Epoch 18, Loss: 0.0010, HR@10: 0.6425, NDCG@10: 0.3660


 63%|██████▎   | 19/30 [31:16<17:32, 95.70s/it]

Epoch 19, Loss: 0.0010, HR@10: 0.6364, NDCG@10: 0.3641


 67%|██████▋   | 20/30 [32:55<16:08, 96.89s/it]

Epoch 20, Loss: 0.0010, HR@10: 0.6411, NDCG@10: 0.3661


 70%|███████   | 21/30 [34:37<14:46, 98.51s/it]

Epoch 21, Loss: 0.0010, HR@10: 0.6349, NDCG@10: 0.3635


 73%|███████▎  | 22/30 [36:20<13:17, 99.63s/it]

Epoch 22, Loss: 0.0010, HR@10: 0.6374, NDCG@10: 0.3651


 77%|███████▋  | 23/30 [37:45<11:07, 95.32s/it]

Epoch 23, Loss: 0.0010, HR@10: 0.6412, NDCG@10: 0.3681


 80%|████████  | 24/30 [39:24<09:39, 96.51s/it]

Epoch 24, Loss: 0.0010, HR@10: 0.6374, NDCG@10: 0.3662


 83%|████████▎ | 25/30 [41:04<08:07, 97.57s/it]

Epoch 25, Loss: 0.0010, HR@10: 0.6384, NDCG@10: 0.3666


 87%|████████▋ | 26/30 [42:40<06:27, 96.98s/it]

Epoch 26, Loss: 0.0010, HR@10: 0.6364, NDCG@10: 0.3662


 90%|█████████ | 27/30 [44:19<04:52, 97.61s/it]

Epoch 27, Loss: 0.0010, HR@10: 0.6382, NDCG@10: 0.3665


 93%|█████████▎| 28/30 [45:34<03:01, 90.85s/it]

Epoch 28, Loss: 0.0010, HR@10: 0.6417, NDCG@10: 0.3677


 97%|█████████▋| 29/30 [47:07<01:31, 91.52s/it]

Epoch 29, Loss: 0.0010, HR@10: 0.6349, NDCG@10: 0.3656


100%|██████████| 30/30 [48:49<00:00, 97.64s/it]

Epoch 30, Loss: 0.0010, HR@10: 0.6409, NDCG@10: 0.3672
Best HR: 0.6425496688741722, Best NDCG: 0.3660134329731621 at Epoch 17


for movielens

HR, NDCG

-   GMF(Factor-8, X-0): 0.6311, 0.3622
-   MLP(Factor-8, X-3): 0.5667, 0.3159
-   NeuMF(Factor-8, X-3): 0.6425, 0.3660

## Reproduce the ablation study --- MLP with different layers

 K: 8
 
 layers (X): 0, 1, 2, 3, 4

In [10]:
model_config_mlp0 = {
    "model_mark": "mlp(mlp_layer=0)",
    "mlp_layers(X)" : 0,
    'mlp_layers': [8],
    'model_type': 'MLP'     #　MLP, NeuMF
}

model_config_mlp1 = {
    "model_mark": "mlp(mlp_layer=1)",
    "mlp_layers(X)" : 1,
    'mlp_layers': [8],
    'model_type': 'MLP'     #　MLP, NeuMF
}

model_config_mlp2 = {
    "model_mark": "mlp(mlp_layer=2)",
    "mlp_layers(X)" : 2,
    'mlp_layers': [16, 8],
    'model_type': 'MLP'     #　MLP, NeuMF
}

model_config_mlp3 = {
    "model_mark": "mlp(mlp_layer=3)",
    "mlp_layers(X)" : 4,
    'mlp_layers': [32, 16, 8],
    'model_type': 'MLP'     #　MLP, NeuMF
}

model_config_mlp4 = {
    "model_mark": "mlp(mlp_layer=4)",
    "mlp_layers(X)" : 4,
    'mlp_layers': [64, 32, 16, 8],
    'model_type': 'MLP'     #　MLP, NeuMF
}

model_train(model_config_mlp0, seed = 42, num_of_negatives=4, num_of_epochs=30)


Using cuda device
Loading preprocessed data from preprocessed_data/ml.pkl
config of model: 
 {'model_mark': 'mlp(mlp_layer=0)', 'mlp_layers(X)': 0, 'mlp_layers': [8], 'model_type': 'MLP', 'num_users': 6040, 'num_items': 3706}


  3%|▎         | 1/30 [00:50<24:25, 50.52s/it]

Epoch 1, Loss: 0.0016, HR@10: 0.4470, NDCG@10: 0.2490


  7%|▋         | 2/30 [01:44<24:25, 52.34s/it]

Epoch 2, Loss: 0.0014, HR@10: 0.4505, NDCG@10: 0.2497


 10%|█         | 3/30 [02:33<22:51, 50.80s/it]

Epoch 3, Loss: 0.0014, HR@10: 0.4513, NDCG@10: 0.2508


 13%|█▎        | 4/30 [03:21<21:31, 49.67s/it]

Epoch 4, Loss: 0.0014, HR@10: 0.4520, NDCG@10: 0.2505


 17%|█▋        | 5/30 [04:15<21:27, 51.49s/it]

Epoch 5, Loss: 0.0014, HR@10: 0.4510, NDCG@10: 0.2506


 20%|██        | 6/30 [05:04<20:15, 50.65s/it]

Epoch 6, Loss: 0.0014, HR@10: 0.4510, NDCG@10: 0.2505


 23%|██▎       | 7/30 [05:57<19:38, 51.24s/it]

Epoch 7, Loss: 0.0014, HR@10: 0.4526, NDCG@10: 0.2526


 27%|██▋       | 8/30 [06:47<18:38, 50.85s/it]

Epoch 8, Loss: 0.0014, HR@10: 0.4520, NDCG@10: 0.2517


 30%|███       | 9/30 [07:38<17:48, 50.86s/it]

Epoch 9, Loss: 0.0014, HR@10: 0.4498, NDCG@10: 0.2518


 33%|███▎      | 10/30 [08:32<17:18, 51.93s/it]

Epoch 10, Loss: 0.0014, HR@10: 0.4500, NDCG@10: 0.2510


 37%|███▋      | 11/30 [09:18<15:54, 50.23s/it]

Epoch 11, Loss: 0.0014, HR@10: 0.4510, NDCG@10: 0.2515


 40%|████      | 12/30 [10:13<15:27, 51.53s/it]

Epoch 12, Loss: 0.0014, HR@10: 0.4503, NDCG@10: 0.2520


 43%|████▎     | 13/30 [11:08<14:54, 52.61s/it]

Epoch 13, Loss: 0.0014, HR@10: 0.4533, NDCG@10: 0.2521


 47%|████▋     | 14/30 [11:57<13:43, 51.46s/it]

Epoch 14, Loss: 0.0014, HR@10: 0.4508, NDCG@10: 0.2520


 50%|█████     | 15/30 [12:50<13:00, 52.03s/it]

Epoch 15, Loss: 0.0014, HR@10: 0.4510, NDCG@10: 0.2520


 53%|█████▎    | 16/30 [13:42<12:09, 52.08s/it]

Epoch 16, Loss: 0.0014, HR@10: 0.4502, NDCG@10: 0.2512


 57%|█████▋    | 17/30 [14:30<10:59, 50.74s/it]

Epoch 17, Loss: 0.0014, HR@10: 0.4495, NDCG@10: 0.2515


 60%|██████    | 18/30 [15:24<10:19, 51.66s/it]

Epoch 18, Loss: 0.0014, HR@10: 0.4507, NDCG@10: 0.2519


 63%|██████▎   | 19/30 [16:14<09:22, 51.10s/it]

Epoch 19, Loss: 0.0014, HR@10: 0.4515, NDCG@10: 0.2519


 67%|██████▋   | 20/30 [17:06<08:36, 51.66s/it]

Epoch 20, Loss: 0.0014, HR@10: 0.4510, NDCG@10: 0.2521


 70%|███████   | 21/30 [18:01<07:51, 52.43s/it]

Epoch 21, Loss: 0.0014, HR@10: 0.4520, NDCG@10: 0.2524


 73%|███████▎  | 22/30 [18:49<06:49, 51.15s/it]

Epoch 22, Loss: 0.0014, HR@10: 0.4522, NDCG@10: 0.2523


 77%|███████▋  | 23/30 [19:22<05:20, 45.83s/it]

Epoch 23, Loss: 0.0014, HR@10: 0.4512, NDCG@10: 0.2517


 80%|████████  | 24/30 [20:16<04:48, 48.10s/it]

Epoch 24, Loss: 0.0014, HR@10: 0.4512, NDCG@10: 0.2522


 83%|████████▎ | 25/30 [21:05<04:01, 48.34s/it]

Epoch 25, Loss: 0.0014, HR@10: 0.4518, NDCG@10: 0.2521


 87%|████████▋ | 26/30 [21:56<03:17, 49.28s/it]

Epoch 26, Loss: 0.0014, HR@10: 0.4505, NDCG@10: 0.2518


 90%|█████████ | 27/30 [22:49<02:31, 50.44s/it]

Epoch 27, Loss: 0.0014, HR@10: 0.4503, NDCG@10: 0.2516


 93%|█████████▎| 28/30 [23:37<01:39, 49.57s/it]

Epoch 28, Loss: 0.0014, HR@10: 0.4497, NDCG@10: 0.2518


 97%|█████████▋| 29/30 [24:28<00:50, 50.18s/it]

Epoch 29, Loss: 0.0014, HR@10: 0.4500, NDCG@10: 0.2519


100%|██████████| 30/30 [25:22<00:00, 50.74s/it]

Epoch 30, Loss: 0.0014, HR@10: 0.4503, NDCG@10: 0.2515
Best HR: 0.45331125827814567, Best NDCG: 0.25205759944116024 at Epoch 12


In [11]:
model_train(model_config_mlp1, seed = 42, num_of_negatives=4, num_of_epochs=30)


Using cuda device
Loading preprocessed data from preprocessed_data/ml.pkl
config of model: 
 {'model_mark': 'mlp(mlp_layer=1)', 'mlp_layers(X)': 1, 'mlp_layers': [8], 'model_type': 'MLP', 'num_users': 6040, 'num_items': 3706}


  3%|▎         | 1/30 [00:48<23:18, 48.23s/it]

Epoch 1, Loss: 0.0016, HR@10: 0.4470, NDCG@10: 0.2490


  7%|▋         | 2/30 [01:42<24:04, 51.60s/it]

Epoch 2, Loss: 0.0014, HR@10: 0.4505, NDCG@10: 0.2497


 10%|█         | 3/30 [02:32<23:03, 51.23s/it]

Epoch 3, Loss: 0.0014, HR@10: 0.4513, NDCG@10: 0.2508


 13%|█▎        | 4/30 [03:18<21:13, 48.99s/it]

Epoch 4, Loss: 0.0014, HR@10: 0.4520, NDCG@10: 0.2505


 17%|█▋        | 5/30 [04:12<21:09, 50.77s/it]

Epoch 5, Loss: 0.0014, HR@10: 0.4510, NDCG@10: 0.2506


 20%|██        | 6/30 [04:59<19:51, 49.63s/it]

Epoch 6, Loss: 0.0014, HR@10: 0.4510, NDCG@10: 0.2505


 23%|██▎       | 7/30 [05:37<17:33, 45.79s/it]

Epoch 7, Loss: 0.0014, HR@10: 0.4526, NDCG@10: 0.2526


 27%|██▋       | 8/30 [06:29<17:31, 47.82s/it]

Epoch 8, Loss: 0.0014, HR@10: 0.4520, NDCG@10: 0.2517


 30%|███       | 9/30 [07:19<16:55, 48.36s/it]

Epoch 9, Loss: 0.0014, HR@10: 0.4498, NDCG@10: 0.2518


 33%|███▎      | 10/30 [08:05<15:53, 47.66s/it]

Epoch 10, Loss: 0.0014, HR@10: 0.4500, NDCG@10: 0.2510


 37%|███▋      | 11/30 [08:59<15:44, 49.73s/it]

Epoch 11, Loss: 0.0014, HR@10: 0.4510, NDCG@10: 0.2515


 40%|████      | 12/30 [09:48<14:46, 49.27s/it]

Epoch 12, Loss: 0.0014, HR@10: 0.4503, NDCG@10: 0.2520


 43%|████▎     | 13/30 [10:41<14:18, 50.49s/it]

Epoch 13, Loss: 0.0014, HR@10: 0.4533, NDCG@10: 0.2521


 47%|████▋     | 14/30 [11:37<13:54, 52.18s/it]

Epoch 14, Loss: 0.0014, HR@10: 0.4508, NDCG@10: 0.2520


 50%|█████     | 15/30 [12:24<12:39, 50.61s/it]

Epoch 15, Loss: 0.0014, HR@10: 0.4510, NDCG@10: 0.2520


 53%|█████▎    | 16/30 [13:16<11:54, 51.03s/it]

Epoch 16, Loss: 0.0014, HR@10: 0.4502, NDCG@10: 0.2512


 57%|█████▋    | 17/30 [14:09<11:12, 51.71s/it]

Epoch 17, Loss: 0.0014, HR@10: 0.4495, NDCG@10: 0.2515


 60%|██████    | 18/30 [14:56<10:01, 50.15s/it]

Epoch 18, Loss: 0.0014, HR@10: 0.4507, NDCG@10: 0.2519


 63%|██████▎   | 19/30 [15:49<09:19, 50.91s/it]

Epoch 19, Loss: 0.0014, HR@10: 0.4515, NDCG@10: 0.2519


 67%|██████▋   | 20/30 [16:40<08:31, 51.15s/it]

Epoch 20, Loss: 0.0014, HR@10: 0.4510, NDCG@10: 0.2521


 70%|███████   | 21/30 [17:27<07:27, 49.71s/it]

Epoch 21, Loss: 0.0014, HR@10: 0.4520, NDCG@10: 0.2524


 73%|███████▎  | 22/30 [18:21<06:48, 51.00s/it]

Epoch 22, Loss: 0.0014, HR@10: 0.4522, NDCG@10: 0.2523


 77%|███████▋  | 23/30 [19:12<05:57, 51.01s/it]

Epoch 23, Loss: 0.0014, HR@10: 0.4512, NDCG@10: 0.2517


 80%|████████  | 24/30 [20:00<05:00, 50.14s/it]

Epoch 24, Loss: 0.0014, HR@10: 0.4512, NDCG@10: 0.2522


 83%|████████▎ | 25/30 [20:53<04:14, 50.93s/it]

Epoch 25, Loss: 0.0014, HR@10: 0.4518, NDCG@10: 0.2521


 87%|████████▋ | 26/30 [21:40<03:19, 49.91s/it]

Epoch 26, Loss: 0.0014, HR@10: 0.4505, NDCG@10: 0.2518


 90%|█████████ | 27/30 [22:32<02:31, 50.54s/it]

Epoch 27, Loss: 0.0014, HR@10: 0.4503, NDCG@10: 0.2516


 93%|█████████▎| 28/30 [23:25<01:42, 51.41s/it]

Epoch 28, Loss: 0.0014, HR@10: 0.4497, NDCG@10: 0.2518


 97%|█████████▋| 29/30 [24:14<00:50, 50.41s/it]

Epoch 29, Loss: 0.0014, HR@10: 0.4500, NDCG@10: 0.2519


100%|██████████| 30/30 [25:06<00:00, 50.22s/it]

Epoch 30, Loss: 0.0014, HR@10: 0.4503, NDCG@10: 0.2515
Best HR: 0.45331125827814567, Best NDCG: 0.25205759944116024 at Epoch 12


In [12]:
model_train(model_config_mlp2, seed = 42, num_of_negatives=4, num_of_epochs=30)


Using cuda device
Loading preprocessed data from preprocessed_data/ml.pkl
config of model: 
 {'model_mark': 'mlp(mlp_layer=2)', 'mlp_layers(X)': 2, 'mlp_layers': [16, 8], 'model_type': 'MLP', 'num_users': 6040, 'num_items': 3706}


  3%|▎         | 1/30 [01:07<32:39, 67.56s/it]

Epoch 1, Loss: 0.0015, HR@10: 0.4523, NDCG@10: 0.2499


  7%|▋         | 2/30 [02:14<31:20, 67.17s/it]

Epoch 2, Loss: 0.0014, HR@10: 0.4508, NDCG@10: 0.2501


 10%|█         | 3/30 [03:13<28:30, 63.36s/it]

Epoch 3, Loss: 0.0014, HR@10: 0.4522, NDCG@10: 0.2521


 13%|█▎        | 4/30 [04:19<27:52, 64.32s/it]

Epoch 4, Loss: 0.0014, HR@10: 0.4545, NDCG@10: 0.2534


 17%|█▋        | 5/30 [05:29<27:44, 66.59s/it]

Epoch 5, Loss: 0.0014, HR@10: 0.4518, NDCG@10: 0.2520


 20%|██        | 6/30 [06:33<26:15, 65.63s/it]

Epoch 6, Loss: 0.0014, HR@10: 0.4556, NDCG@10: 0.2551


 23%|██▎       | 7/30 [07:44<25:52, 67.52s/it]

Epoch 7, Loss: 0.0014, HR@10: 0.4573, NDCG@10: 0.2549


 27%|██▋       | 8/30 [08:50<24:35, 67.05s/it]

Epoch 8, Loss: 0.0014, HR@10: 0.4598, NDCG@10: 0.2579


 30%|███       | 9/30 [10:01<23:52, 68.21s/it]

Epoch 9, Loss: 0.0014, HR@10: 0.4634, NDCG@10: 0.2588


 33%|███▎      | 10/30 [11:08<22:35, 67.80s/it]

Epoch 10, Loss: 0.0014, HR@10: 0.4662, NDCG@10: 0.2601


 37%|███▋      | 11/30 [12:06<20:29, 64.72s/it]

Epoch 11, Loss: 0.0013, HR@10: 0.4709, NDCG@10: 0.2624


 40%|████      | 12/30 [13:13<19:36, 65.37s/it]

Epoch 12, Loss: 0.0013, HR@10: 0.4775, NDCG@10: 0.2651


 43%|████▎     | 13/30 [14:23<18:57, 66.90s/it]

Epoch 13, Loss: 0.0013, HR@10: 0.4803, NDCG@10: 0.2667


 47%|████▋     | 14/30 [15:32<18:01, 67.58s/it]

Epoch 14, Loss: 0.0013, HR@10: 0.4820, NDCG@10: 0.2675


 50%|█████     | 15/30 [16:44<17:11, 68.74s/it]

Epoch 15, Loss: 0.0013, HR@10: 0.4854, NDCG@10: 0.2704


 53%|█████▎    | 16/30 [17:54<16:08, 69.17s/it]

Epoch 16, Loss: 0.0013, HR@10: 0.4904, NDCG@10: 0.2714


 57%|█████▋    | 17/30 [19:03<15:01, 69.31s/it]

Epoch 17, Loss: 0.0013, HR@10: 0.4916, NDCG@10: 0.2740


 60%|██████    | 18/30 [20:15<14:01, 70.12s/it]

Epoch 18, Loss: 0.0013, HR@10: 0.4967, NDCG@10: 0.2748


 63%|██████▎   | 19/30 [21:24<12:47, 69.78s/it]

Epoch 19, Loss: 0.0013, HR@10: 0.4972, NDCG@10: 0.2752


 67%|██████▋   | 20/30 [22:38<11:47, 70.80s/it]

Epoch 20, Loss: 0.0013, HR@10: 0.5007, NDCG@10: 0.2775


 70%|███████   | 21/30 [23:47<10:32, 70.32s/it]

Epoch 21, Loss: 0.0013, HR@10: 0.5010, NDCG@10: 0.2798


 73%|███████▎  | 22/30 [25:00<09:30, 71.26s/it]

Epoch 22, Loss: 0.0013, HR@10: 0.5028, NDCG@10: 0.2785


 77%|███████▋  | 23/30 [26:01<07:56, 68.11s/it]

Epoch 23, Loss: 0.0013, HR@10: 0.5022, NDCG@10: 0.2794


 80%|████████  | 24/30 [26:47<06:08, 61.47s/it]

Epoch 24, Loss: 0.0013, HR@10: 0.5040, NDCG@10: 0.2816


 83%|████████▎ | 25/30 [27:31<04:41, 56.34s/it]

Epoch 25, Loss: 0.0013, HR@10: 0.5070, NDCG@10: 0.2828


 87%|████████▋ | 26/30 [28:07<03:20, 50.19s/it]

Epoch 26, Loss: 0.0013, HR@10: 0.5093, NDCG@10: 0.2838


 90%|█████████ | 27/30 [29:02<02:34, 51.66s/it]

Epoch 27, Loss: 0.0013, HR@10: 0.5060, NDCG@10: 0.2837


 93%|█████████▎| 28/30 [30:01<01:47, 53.88s/it]

Epoch 28, Loss: 0.0013, HR@10: 0.5053, NDCG@10: 0.2826


 97%|█████████▋| 29/30 [31:03<00:56, 56.15s/it]

Epoch 29, Loss: 0.0013, HR@10: 0.5104, NDCG@10: 0.2847


100%|██████████| 30/30 [32:04<00:00, 64.17s/it]

Epoch 30, Loss: 0.0013, HR@10: 0.5103, NDCG@10: 0.2848
Best HR: 0.510430463576159, Best NDCG: 0.28466994624063796 at Epoch 28


In [13]:
model_train(model_config_mlp3, seed = 42, num_of_negatives=4, num_of_epochs=30)


Using cuda device
Loading preprocessed data from preprocessed_data/ml.pkl
config of model: 
 {'model_mark': 'mlp(mlp_layer=3)', 'mlp_layers(X)': 4, 'mlp_layers': [32, 16, 8], 'model_type': 'MLP', 'num_users': 6040, 'num_items': 3706}


  3%|▎         | 1/30 [00:35<17:11, 35.56s/it]

Epoch 1, Loss: 0.0015, HR@10: 0.4434, NDCG@10: 0.2497


  7%|▋         | 2/30 [01:38<24:04, 51.58s/it]

Epoch 2, Loss: 0.0014, HR@10: 0.4515, NDCG@10: 0.2518


 10%|█         | 3/30 [02:59<29:18, 65.13s/it]

Epoch 3, Loss: 0.0014, HR@10: 0.4545, NDCG@10: 0.2537


 13%|█▎        | 4/30 [04:25<31:48, 73.40s/it]

Epoch 4, Loss: 0.0014, HR@10: 0.4624, NDCG@10: 0.2574


 17%|█▋        | 5/30 [05:49<32:13, 77.33s/it]

Epoch 5, Loss: 0.0013, HR@10: 0.4829, NDCG@10: 0.2651


 20%|██        | 6/30 [07:12<31:36, 79.00s/it]

Epoch 6, Loss: 0.0013, HR@10: 0.4911, NDCG@10: 0.2717


 23%|██▎       | 7/30 [08:35<30:45, 80.26s/it]

Epoch 7, Loss: 0.0013, HR@10: 0.5005, NDCG@10: 0.2741


 27%|██▋       | 8/30 [09:56<29:35, 80.72s/it]

Epoch 8, Loss: 0.0013, HR@10: 0.5070, NDCG@10: 0.2787


 30%|███       | 9/30 [11:14<27:54, 79.75s/it]

Epoch 9, Loss: 0.0013, HR@10: 0.5132, NDCG@10: 0.2842


 33%|███▎      | 10/30 [12:34<26:39, 79.97s/it]

Epoch 10, Loss: 0.0012, HR@10: 0.5129, NDCG@10: 0.2839


 37%|███▋      | 11/30 [14:01<25:55, 81.87s/it]

Epoch 11, Loss: 0.0012, HR@10: 0.5242, NDCG@10: 0.2899


 40%|████      | 12/30 [15:27<24:57, 83.21s/it]

Epoch 12, Loss: 0.0012, HR@10: 0.5263, NDCG@10: 0.2924


 43%|████▎     | 13/30 [16:49<23:29, 82.89s/it]

Epoch 13, Loss: 0.0012, HR@10: 0.5275, NDCG@10: 0.2976


 47%|████▋     | 14/30 [18:15<22:19, 83.71s/it]

Epoch 14, Loss: 0.0012, HR@10: 0.5301, NDCG@10: 0.2975


 50%|█████     | 15/30 [19:36<20:45, 83.02s/it]

Epoch 15, Loss: 0.0012, HR@10: 0.5412, NDCG@10: 0.3031


 53%|█████▎    | 16/30 [21:04<19:44, 84.58s/it]

Epoch 16, Loss: 0.0012, HR@10: 0.5421, NDCG@10: 0.3062


 57%|█████▋    | 17/30 [22:30<18:23, 84.92s/it]

Epoch 17, Loss: 0.0012, HR@10: 0.5439, NDCG@10: 0.3069


 60%|██████    | 18/30 [23:54<16:54, 84.52s/it]

Epoch 18, Loss: 0.0012, HR@10: 0.5478, NDCG@10: 0.3064


 63%|██████▎   | 19/30 [25:20<15:36, 85.10s/it]

Epoch 19, Loss: 0.0012, HR@10: 0.5588, NDCG@10: 0.3112


 67%|██████▋   | 20/30 [26:40<13:57, 83.73s/it]

Epoch 20, Loss: 0.0012, HR@10: 0.5533, NDCG@10: 0.3090


 70%|███████   | 21/30 [27:29<10:58, 73.17s/it]

Epoch 21, Loss: 0.0012, HR@10: 0.5518, NDCG@10: 0.3075


 73%|███████▎  | 22/30 [28:50<10:04, 75.51s/it]

Epoch 22, Loss: 0.0012, HR@10: 0.5579, NDCG@10: 0.3101


 77%|███████▋  | 23/30 [30:15<09:08, 78.40s/it]

Epoch 23, Loss: 0.0012, HR@10: 0.5616, NDCG@10: 0.3128


 80%|████████  | 24/30 [31:36<07:54, 79.03s/it]

Epoch 24, Loss: 0.0012, HR@10: 0.5632, NDCG@10: 0.3139


 83%|████████▎ | 25/30 [33:02<06:46, 81.27s/it]

Epoch 25, Loss: 0.0012, HR@10: 0.5667, NDCG@10: 0.3159


 87%|████████▋ | 26/30 [34:27<05:29, 82.43s/it]

Epoch 26, Loss: 0.0012, HR@10: 0.5637, NDCG@10: 0.3134


 90%|█████████ | 27/30 [35:49<04:06, 82.18s/it]

Epoch 27, Loss: 0.0012, HR@10: 0.5649, NDCG@10: 0.3134


 93%|█████████▎| 28/30 [37:13<02:45, 82.79s/it]

Epoch 28, Loss: 0.0011, HR@10: 0.5644, NDCG@10: 0.3135


 97%|█████████▋| 29/30 [38:34<01:22, 82.10s/it]

Epoch 29, Loss: 0.0011, HR@10: 0.5581, NDCG@10: 0.3146


100%|██████████| 30/30 [39:58<00:00, 79.96s/it]

Epoch 30, Loss: 0.0011, HR@10: 0.5598, NDCG@10: 0.3126
Best HR: 0.5667218543046357, Best NDCG: 0.3158625751324939 at Epoch 24


In [14]:
model_train(model_config_mlp4, seed = 42, num_of_negatives=4, num_of_epochs=30)


Using cuda device
Loading preprocessed data from preprocessed_data/ml.pkl
config of model: 
 {'model_mark': 'mlp(mlp_layer=4)', 'mlp_layers(X)': 4, 'mlp_layers': [64, 32, 16, 8], 'model_type': 'MLP', 'num_users': 6040, 'num_items': 3706}


  3%|▎         | 1/30 [01:38<47:50, 98.99s/it]

Epoch 1, Loss: 0.0014, HR@10: 0.4417, NDCG@10: 0.2477


  7%|▋         | 2/30 [03:14<45:12, 96.88s/it]

Epoch 2, Loss: 0.0014, HR@10: 0.4543, NDCG@10: 0.2520


 10%|█         | 3/30 [04:55<44:26, 98.76s/it]

Epoch 3, Loss: 0.0014, HR@10: 0.4666, NDCG@10: 0.2583


 13%|█▎        | 4/30 [06:33<42:45, 98.69s/it]

Epoch 4, Loss: 0.0013, HR@10: 0.4909, NDCG@10: 0.2705


 17%|█▋        | 5/30 [08:12<41:03, 98.53s/it]

Epoch 5, Loss: 0.0013, HR@10: 0.5048, NDCG@10: 0.2794


 20%|██        | 6/30 [09:51<39:33, 98.91s/it]

Epoch 6, Loss: 0.0012, HR@10: 0.5195, NDCG@10: 0.2875


 23%|██▎       | 7/30 [11:26<37:21, 97.46s/it]

Epoch 7, Loss: 0.0012, HR@10: 0.5300, NDCG@10: 0.2955


 27%|██▋       | 8/30 [13:08<36:16, 98.91s/it]

Epoch 8, Loss: 0.0012, HR@10: 0.5462, NDCG@10: 0.3037


 30%|███       | 9/30 [14:47<34:40, 99.09s/it]

Epoch 9, Loss: 0.0012, HR@10: 0.5593, NDCG@10: 0.3140


 33%|███▎      | 10/30 [16:24<32:47, 98.37s/it]

Epoch 10, Loss: 0.0012, HR@10: 0.5623, NDCG@10: 0.3178


 37%|███▋      | 11/30 [18:06<31:28, 99.37s/it]

Epoch 11, Loss: 0.0011, HR@10: 0.5733, NDCG@10: 0.3203


 40%|████      | 12/30 [19:44<29:42, 99.02s/it]

Epoch 12, Loss: 0.0011, HR@10: 0.5714, NDCG@10: 0.3223


 43%|████▎     | 13/30 [21:26<28:20, 100.04s/it]

Epoch 13, Loss: 0.0011, HR@10: 0.5758, NDCG@10: 0.3241


 47%|████▋     | 14/30 [23:05<26:33, 99.58s/it] 

Epoch 14, Loss: 0.0011, HR@10: 0.5791, NDCG@10: 0.3264


 50%|█████     | 15/30 [24:42<24:44, 98.96s/it]

Epoch 15, Loss: 0.0011, HR@10: 0.5806, NDCG@10: 0.3297


 53%|█████▎    | 16/30 [26:25<23:18, 99.92s/it]

Epoch 16, Loss: 0.0011, HR@10: 0.5858, NDCG@10: 0.3308


 57%|█████▋    | 17/30 [28:04<21:35, 99.66s/it]

Epoch 17, Loss: 0.0011, HR@10: 0.5922, NDCG@10: 0.3350


 60%|██████    | 18/30 [29:43<19:53, 99.47s/it]

Epoch 18, Loss: 0.0011, HR@10: 0.5937, NDCG@10: 0.3359


 63%|██████▎   | 19/30 [31:24<18:20, 100.07s/it]

Epoch 19, Loss: 0.0011, HR@10: 0.5952, NDCG@10: 0.3382


 67%|██████▋   | 20/30 [33:03<16:38, 99.85s/it] 

Epoch 20, Loss: 0.0011, HR@10: 0.5967, NDCG@10: 0.3363


 70%|███████   | 21/30 [34:43<14:58, 99.88s/it]

Epoch 21, Loss: 0.0010, HR@10: 0.6025, NDCG@10: 0.3391


 73%|███████▎  | 22/30 [36:23<13:18, 99.82s/it]

Epoch 22, Loss: 0.0010, HR@10: 0.6053, NDCG@10: 0.3423


 77%|███████▋  | 23/30 [37:59<11:31, 98.75s/it]

Epoch 23, Loss: 0.0010, HR@10: 0.6003, NDCG@10: 0.3374


 80%|████████  | 24/30 [39:41<09:57, 99.57s/it]

Epoch 24, Loss: 0.0010, HR@10: 0.6012, NDCG@10: 0.3394


 83%|████████▎ | 25/30 [41:21<08:18, 99.64s/it]

Epoch 25, Loss: 0.0010, HR@10: 0.6050, NDCG@10: 0.3431


 87%|████████▋ | 26/30 [42:58<06:36, 99.06s/it]

Epoch 26, Loss: 0.0010, HR@10: 0.6003, NDCG@10: 0.3407


 90%|█████████ | 27/30 [44:40<04:59, 99.84s/it]

Epoch 27, Loss: 0.0010, HR@10: 0.6017, NDCG@10: 0.3409


 93%|█████████▎| 28/30 [46:18<03:18, 99.38s/it]

Epoch 28, Loss: 0.0010, HR@10: 0.6002, NDCG@10: 0.3411


 97%|█████████▋| 29/30 [47:57<01:39, 99.25s/it]

Epoch 29, Loss: 0.0010, HR@10: 0.5977, NDCG@10: 0.3404


100%|██████████| 30/30 [49:37<00:00, 99.23s/it]

Epoch 30, Loss: 0.0010, HR@10: 0.5982, NDCG@10: 0.3419
Best HR: 0.6052980132450331, Best NDCG: 0.3423428631718419 at Epoch 21


HR, NDCG

-   MLP0: 0.4533, 0.2521
-   MLP1: 0.4533, 0.2521
-   MLP2: 0.5104, 0.2847
-   MLP3: 0.5667, 0.3159
-   MLP4: 0.6053, 0.3423